multimodal/src/autogluon/multimodal/matcher.py

```diff
@@ -1173,33 +1329,64 @@ def semantic_search(
+            query_embeddings = self.extract_embedding(
+                query_data[query_start_idx : query_start_idx + query_chunk_size],
+                signature=QUERY,
+                id_mappings=id_mappings,
+                as_tensor=True,
+            )
+            # Iterate over chunks of the corpus
+            for response_start_idx in range(0, len(response_data), response_chunk_size):
+                response_embeddings = self.extract_embedding(
+                    response_data[response_start_idx : response_start_idx + response_chunk_size],
+                    signature=RESPONSE,
+                    id_mappings=id_mappings,
+                    as_tensor=True,
+                )
+                # Compute cosine similarities
+                scores = compute_semantic_similarity(
+                    a=query_embeddings,
+                    b=response_embeddings,
+                    cosine=cosine,
+                )
 
-        if (
-            as_pandas is None and isinstance(query_data, pd.DataFrame) and isinstance(response_data, pd.DataFrame)
-        ) or as_pandas is True:
-            ret = self._as_pandas(data=query_data, to_be_converted=ret)
+                # Get top-k scores
+                scores_top_k_values, scores_top_k_idx = torch.topk(
+                    scores,
+                    k=min(top_k, len(scores[0])),
+                    dim=1,
+                    largest=True,
+                    sorted=False,
+                )
+                scores_top_k_values = scores_top_k_values.cpu().tolist()
+                scores_top_k_idx = scores_top_k_idx.cpu().tolist()
+
+                for query_itr in range(len(scores)):
+                    for sub_response_id, score in zip(scores_top_k_idx[query_itr], scores_top_k_values[query_itr]):
+                        corpus_id = response_start_idx + sub_response_id
+                        query_id = query_start_idx + query_itr
+                        if len(queries_result_list[query_id]) < top_k:
+                            heapq.heappush(
+                                queries_result_list[query_id], (score, corpus_id)
+                            )  # heaqp tracks the quantity of the first element in the tuple
+                        else:
+                            heapq.heappushpop(queries_result_list[query_id], (score, corpus_id))
+
+        # change the data format and sort
+        for query_id in range(len(queries_result_list)):
+            for doc_itr in range(len(queries_result_list[query_id])):
+                score, corpus_id = queries_result_list[query_id][doc_itr]
+                queries_result_list[query_id][doc_itr] = {"corpus_id": corpus_id, "score": score}
+            queries_result_list[query_id] = sorted(
+                queries_result_list[query_id], key=lambda x: x["score"], reverse=True
+            )
 
-        return ret
+        return queries_result_list
```

Comment:
We should not add semantic search here.

Reply:
Done. Moved `semantic_search` into utils.